docs/docs/creating-a-source-plugin.md

```diff
@@ -18,147 +18,426 @@ At a high-level, a source plugin:
+- **Add event-based sync**. Some data sources keep event logs and are able to return a list of objects modified since a given time. If you're building a source plugin, you can store the last time you fetched data using the [cache](/docs/creating-a-source-plugin/#caching-data-between-runs) or [`setPluginStatus`](/docs/actions/#setPluginStatus) and then only sync down nodes that have been modified since that time. [gatsby-source-contentful](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-source-contentful) is an example of a source plugin that does this.
+
+If possible, the proactive listener approach creates the best experience if existing APIs in the data source can support it (or you have access to build support into the data source).
+
+The code to support this behavior looks like this:
+
+```javascript:title=source-plugin/gatsby-node.js
+exports.sourceNodes = async ({ actions }, pluginOptions) => {
+  const { createNode, touchNode, deleteNode } = actions
+
+  // highlight-start
+  // touch nodes to ensure they aren't garbage collected
+  getNodesByType(`YourSourceType`).forEach(node => touchNode({ nodeId: node.id }))
+
+  // ensure a plugin is in a preview mode and/or supports listening
+  if (pluginOptions.preview) {
+    const subscription = await subscription(SUBSCRIPTION_TO_WEBSOCKET)
+    subscription.subscribe(({ data: newData }) => {
+      newData.forEach(newDatum => {
+        switch (newDatum.status) {
+          case "deleted":
+            deleteNode({
+              node: getNode(createNodeId(`YourSourceType-${newDatum.uuid}`)),
+            })
+            break
+          case "created":
+          case "updated":
+          default:
+            // created and updated can be handled by the same code path
+            // the post's id is presumed to stay constant (or can be inferred)
+            createNode(processDatum(newDatum))
+            break
+        )
+      }
+    })
+  }
+  // highlight-end
+
+  const data = await client.query(QUERY_TO_API)
+
+  // Process data and create nodes.using a custom processDatum function
+  data.forEach(datum => createNode(processDatum(datum)))
+
+  // You're done, return.
+  return
+}
+```
+
+_Note: This is pseudo code to illustrate the logic and concept of how these plugins function, you can see an example in the [creating source plugins](https://github.com/gatsbyjs/gatsby/tree/master/examples/creating-source-plugins) repository._
+
+Because the code in sourceNodes will be reinvoked when changes in the data source occur, a few steps need to be taken to make sure that Gatsby is tracking the existing nodes that have already been sourced along with the new data. A first step is ensuring that the existing nodes created are not garbage collected which is done by "touching" the nodes with the [`touchNode` action](/docs/actions/#touchNode).
```

Comment:
```suggestion
Because the code in `sourceNodes` is reinvoked when changes in the data source occur, a few steps need to be taken to ensure that Gatsby is tracking the existing nodes as well as the new data. A first step is ensuring that the existing nodes created are not garbage collected which is done by "touching" the nodes with the [`touchNode` action](/docs/actions/#touchNode).
```